File: src/Heap.hs
module Heap where

class Heap h where
  empty :: h a
  isEmpty :: h a -> Bool
  insert :: Ord a => a -> h a -> h a
  merge :: Ord a => h a -> h a -> h a
  findMin :: h a -> Maybe a
  deleteMin :: Ord a => h a -> Maybe (h a)

data HeapT a = E | T Int a (HeapT a) (HeapT a) deriving (Show, Eq)

rank :: HeapT a -> Int
rank E = 0
rank (T r _ _ _) = r

makeT :: a -> HeapT a -> HeapT a -> HeapT a
makeT x a b =
  if rank a >= rank b
    then T (rank b + 1) x a b
    else T (rank a + 1) x b a

insert2 :: Ord a => a -> HeapT a -> HeapT a
insert2 x E = T 1 x E E
-- 任意の左の子のランクが、右の子のランクと同じかそれ 以上である
-- ノードのランクは、そのノードの右スパイン
-- (right spine、そのノードから空ノードまでの最も右の経路)の長さとして定義される
-- x >= y の場合: 右に挿入する
-- x < y の場合: xを新しい根にしてyを左に挿入する
insert2 x (T _ y a b) = if x >= y then makeT y a (insert2 x b)
                                  else makeT x (insert2 y a) b


instance Heap HeapT where
  empty = E

  isEmpty E = True
  isEmpty _ = False

  merge E E = E
  merge h@T {} E = h
  merge E h@T {} = h
  merge h1@(T _ x a1 b1) h2@(T _ y a2 b2) =
    if x <= y
      then makeT x a1 (merge b1 h2)
      else makeT y a2 (merge h1 b2)

  insert x h = merge (T 1 x E E) h

  findMin E = Nothing
  findMin (T _ x a b) = Just x

  deleteMin E = Nothing
  deleteMin (T _ x a b) = Just $ merge a b

File: scala/Heap.scala
trait Heap[F[_]] {
  def empty[A]: F[A]
  def isEmpty[A](h: F[A]): Boolean
  def findMin[A: Ordering](h: F[A]): Option[A]
  def insert[A: Ordering](a: A, h: F[A]): F[A]
  def merge[A: Ordering](h1: F[A], h2: F[A]): F[A]
  def deleteMin[A: Ordering](h: F[A]): Option[(F[A], A)]
}

object Heap {
  def apply[F[_]](implicit h: Heap[F]): Heap[F] = h
  implicit class HeapOps[F[_]: Heap, A: Ordering](h: F[A]) {
    def insert(a: A) = Heap[F].insert(a, h)
    def deleteMin = Heap[F].deleteMin(h)
    def findMin = Heap[F].findMin(h)
    def merge(h2: F[A]): F[A] = Heap[F].merge(h, h2)
  }
}
